test/e2e-helm/e2e_suite.go

```diff
@@ -0,0 +1,211 @@
+					true,
+					"pods", controllerPodName, "-o", "jsonpath={.status.phase}")
+				Expect(err).NotTo(HaveOccurred())
+				if status != "Running" {
+					return fmt.Errorf("controller pod in %s status", status)
+				}
+				return nil
+			}
+			Eventually(verifyControllerUp, time.Minute, time.Second).Should(Succeed())
+
+			By("granting permissions to access the metrics and read the token")
+			_, err = tc.Kubectl.Command(
+				"create", "clusterrolebinding", fmt.Sprintf("metrics-%s", tc.TestSuffix),
+				fmt.Sprintf("--clusterrole=e2e-%s-metrics-reader", tc.TestSuffix),
+				fmt.Sprintf("--serviceaccount=%s:default", tc.Kubectl.Namespace))
+			Expect(err).NotTo(HaveOccurred())
+
+			By("getting the token")
+			b64Token, err := tc.Kubectl.Get(true, "secrets", "-o=jsonpath={.items[0].data.token}")
+			Expect(err).NotTo(HaveOccurred())
+			token, err := base64.StdEncoding.DecodeString(strings.TrimSpace(b64Token))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(token)).To(BeNumerically(">", 0))
+
+			By("creating a pod with curl image")
+			cmdOpts := []string{
+				"run", "--generator=run-pod/v1", "curl", "--image=curlimages/curl:7.68.0", "--restart=OnFailure", "--",
+				"curl", "-v", "-k", "-H", fmt.Sprintf(`Authorization: Bearer %s`, token),
+				fmt.Sprintf("https://e2e-%v-controller-manager-metrics-service.e2e-%v-system.svc:8443/metrics",
+					tc.TestSuffix, tc.TestSuffix),
+			}
+			_, err = tc.Kubectl.CommandInNamespace(cmdOpts...)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("validating the curl pod running as expected")
+			verifyCurlUp := func() error {
+				// Validate pod status
+				status, err := tc.Kubectl.Get(
+					true,
+					"pods", "curl", "-o", "jsonpath={.status.phase}")
+				Expect(err).NotTo(HaveOccurred())
+				if status != "Completed" && status != "Succeeded" {
+					return fmt.Errorf("curl pod in %s status", status)
+				}
+				return nil
+			}
+			Eventually(verifyCurlUp, 30*time.Second, time.Second).Should(Succeed())
+
+			By("validating the metrics endpoint is serving as expected")
+			getCurlLogs := func() string {
+				logOutput, err := tc.Kubectl.Logs("curl")
+				Expect(err).NotTo(HaveOccurred())
+				return logOutput
+			}
+			Eventually(getCurlLogs, 10*time.Second, time.Second).Should(ContainSubstring("< HTTP/2 200"))
+
+			By("validate Service Monitor for Prometheus is applied in the namespace")
+			_, err = tc.Kubectl.Get(
+				true,
+				"ServiceMonitor")
+			Expect(err).NotTo(HaveOccurred())
+
+			By("creating an instance of CR in another namespace")
+			// currently controller-runtime doesn't provide a readiness probe, we retry a few times
+			// we can change it to probe the readiness endpoint after CR supports it.
+			sampleFile := filepath.Join("config", "samples",
+				fmt.Sprintf("%s_%s_%s.yaml", tc.Group, tc.Version, strings.ToLower(tc.Kind)))
+			Eventually(func() error {
+				_, err = tc.Kubectl.Apply(false, "-f", sampleFile)
+				return err
+			}, time.Minute, time.Second).Should(Succeed())
```

Comment:
Do we need this wrapped in `Eventually`? The only reason this should fail is if the CRD does not exist, but it should by this point because we ran `make install`.

Do we ever expect this to fail, even once?